File: backend/src/utils/validators.js
const { body, param, query, validationResult } = require('express-validator');

const handleValidationErrors = (req, res, next) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return res.status(400).json({
      success: false,
      message: 'Validation failed',
      errors: errors.array().map((e) => ({ field: e.path, message: e.msg })),
    });
  }
  next();
};

const authValidators = {
  verifyOtp: [
    body('phone').notEmpty().withMessage('Phone number is required').isMobilePhone().withMessage('Invalid phone number'),
    body('firebaseToken').notEmpty().withMessage('Firebase token is required'),
    handleValidationErrors,
  ],
  refreshToken: [body('refreshToken').notEmpty().withMessage('Refresh token is required'), handleValidationErrors],
};

const customerValidators = {
  create: [
    body('name')
      .notEmpty()
      .withMessage('Customer name is required')
      .isLength({ min: 2, max: 100 })
      .withMessage('Name must be 2-100 characters'),
    body('phone')
      .notEmpty()
      .withMessage('Phone number is required')
      .matches(/^[6-9]\d{9}$/)
      .withMessage('Invalid Indian mobile number'),
    body('email').optional().isEmail().withMessage('Invalid email address'),
    handleValidationErrors,
  ],
  update: [
    param('id').isUUID().withMessage('Invalid customer ID'),
    body('name').optional().isLength({ min: 2, max: 100 }).withMessage('Name must be 2-100 characters'),
    body('phone').optional().matches(/^[6-9]\d{9}$/).withMessage('Invalid Indian mobile number'),
    body('email').optional().isEmail().withMessage('Invalid email address'),
    handleValidationErrors,
  ],
  getById: [param('id').isUUID().withMessage('Invalid customer ID'), handleValidationErrors],
  search: [
    query('q').optional().isString().isLength({ max: 100 }),
    query('page').optional().isInt({ min: 1 }).withMessage('Page must be a positive integer'),
    query('limit').optional().isInt({ min: 1, max: 100 }).withMessage('Limit must be between 1-100'),
    handleValidationErrors,
  ],
};

const transactionValidators = {
  create: [
    body('customerId').notEmpty().withMessage('Customer ID is required').isUUID().withMessage('Invalid customer ID'),
    body('amount')
      .notEmpty()
      .withMessage('Amount is required')
      .isFloat({ min: 0.01 })
      .withMessage('Amount must be greater than 0'),
    body('type')
      .notEmpty()
      .withMessage('Transaction type is required')
      .isIn(['credit', 'debit'])
      .withMessage('Type must be credit or debit'),
    body('notes').optional().isLength({ max: 500 }).withMessage('Notes cannot exceed 500 characters'),
    handleValidationErrors,
  ],
  getByCustomer: [
    param('customerId').isUUID().withMessage('Invalid customer ID'),
    query('page').optional().isInt({ min: 1 }),
    query('limit').optional().isInt({ min: 1, max: 100 }),
    query('startDate').optional().isISO8601().withMessage('Invalid start date'),
    query('endDate').optional().isISO8601().withMessage('Invalid end date'),
    handleValidationErrors,
  ],
};

const paymentValidators = {
  initiate: [
    body('customerId').notEmpty().isUUID().withMessage('Invalid customer ID'),
    body('amount').notEmpty().isFloat({ min: 1 }).withMessage('Amount must be at least 1'),
    body('upiId').optional().isString().withMessage('Invalid UPI ID'),
    handleValidationErrors,
  ],
  verify: [
    body('razorpayPaymentId').notEmpty().withMessage('Payment ID is required'),
    body('razorpayOrderId').notEmpty().withMessage('Order ID is required'),
    body('razorpaySignature').notEmpty().withMessage('Signature is required'),
    handleValidationErrors,
  ],
};

const reminderValidators = {
  send: [
    body('customerId').notEmpty().isUUID().withMessage('Invalid customer ID'),
    body('message').notEmpty().isLength({ min: 10, max: 500 }).withMessage('Message must be 10-500 characters'),
    handleValidationErrors,
  ],
  schedule: [
    body('customerId').notEmpty().isUUID().withMessage('Invalid customer ID'),
    body('message').notEmpty().isLength({ min: 10, max: 500 }),
    body('scheduledAt').notEmpty().isISO8601().withMessage('Invalid scheduled date'),
    handleValidationErrors,
  ],
};

module.exports = {
  authValidators,
  customerValidators,
  transactionValidators,
  paymentValidators,
  reminderValidators,
  handleValidationErrors,
};

File: lib/utils/validators.dart
class AppValidators {
  static String? validatePhone(String? value) {
    if (value == null || value.trim().isEmpty) {
      return 'Phone number is required';
    }
    final clean = value.replaceAll(RegExp(r'[^0-9]'), '');
    if (clean.length != 10) {
      return 'Enter a valid 10-digit phone number';
    }
    if (!RegExp(r'^[6-9]\d{9}$').hasMatch(clean)) {
      return 'Enter a valid Indian mobile number';
    }
    return null;
  }

  static String? validateOtp(String? value) {
    if (value == null || value.trim().isEmpty) return 'OTP is required';
    if (value.trim().length != 6) return 'Enter a valid 6-digit OTP';
    if (!RegExp(r'^\d{6}$').hasMatch(value.trim())) return 'OTP must be 6 digits';
    return null;
  }

  static String? validateName(String? value) {
    if (value == null || value.trim().isEmpty) return 'Name is required';
    if (value.trim().length < 2) return 'Name must be at least 2 characters';
    if (value.trim().length > 100) return 'Name is too long';
    return null;
  }

  static String? validateEmail(String? value) {
    if (value == null || value.trim().isEmpty) return null; // optional
    if (!RegExp(r'^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$').hasMatch(value.trim())) {
      return 'Enter a valid email address';
    }
    return null;
  }

  static String? validateAmount(String? value) {
    if (value == null || value.trim().isEmpty) return 'Amount is required';
    final amount = double.tryParse(value.replaceAll(',', ''));
    if (amount == null) return 'Enter a valid amount';
    if (amount <= 0) return 'Amount must be greater than 0';
    if (amount > 10000000) return 'Amount cannot exceed ₹1,00,00,000';
    return null;
  }

  static String? validateRequired(String? value, {String? fieldName}) {
    if (value == null || value.trim().isEmpty) {
      return '${fieldName ?? 'This field'} is required';
    }
    return null;
  }

  static String? validateNotes(String? value) {
    if (value != null && value.length > 500) {
      return 'Notes cannot exceed 500 characters';
    }
    return null;
  }

  static String? validateUpiId(String? value) {
    if (value == null || value.trim().isEmpty) return null; // optional
    if (!RegExp(r'^[\w.-]+@[\w]+$').hasMatch(value.trim())) {
      return 'Enter a valid UPI ID (e.g., name@upi)';
    }
    return null;
  }
}
